### DISCUSSION TITLE
`vim.print()` vs `print(vim.inspect())`

### DISCUSSION THREAD

**Discussion by UnkwUsr** [neovim/neovim#24810] [General]
I see that in async context `vim.print()` does [stub](https://github.com/neovim/neovim/blob/master/runtime/lua/vim/_editor.lua#L879-L882C6) and prints only `table: 0x7f6d1deaa7a0` because it uses `nvim_out_write` in its implementation (which can't be called in async context), while `print(vim.inspect())` still works in async context.

Other difference I see is with strings (only strings itself. Tables with any strings will be printed identically by both functions):

```
> vim.print("a\n\nb")
a

b
> print(vim.inspect("a\n\nb"))
"a\n\nb"
```

I tried printing global object `package` with both functions and output was identical.

And as a "pro" for `vim.print()`: it accepts multiple arguments, while `vim.inspect()` accepts only 1.

Also `vim.print()` additionally returns all arguments untouched.

(Any more differences?)

So my questions are:

1. Can/should we make `vim.print()` work in async context? (then issue should be raised)
2. Is it safe to use `print(vim.inspect())` everywhere, especially in async contexts? (I have `function P(a) print(vim.inspect(a)) return a end` in my config)

**Comment by justinmk** [comment]
> 1. Can/should we make `vim.print()` work in async context?

sure

> 2\. Is it safe to use `print(vim.inspect())` everywhere, especially in async contexts?

yes
